### PR TITLE
feat(kernels): add Triton batch-invariant attention

### LINKED ISSUE #147: [WS1][kernels] Batch-invariant attention (standard softmax)
Part of WS1 — Full Batch-Invariant Forward Chain (epic: #<WS1 tracking issue>)

## Why

Flash-style attention parallelizes over the KV dimension and merges partial results, and the number of KV splits depends on sequence length and batch — so the online-softmax accumulation order changes with batch configuration, breaking invariance. Because attention mixes information across positions, even tiny drift here spreads to every later token.

## Scope

Provide a batch-invariant standard-softmax attention for the forward chain.

- Start from a deterministic masked-softmax + logsumexp (LSE) kernel with a fixed max / sum-exp order, then extend to the full attention output (P x V).
- Avoid split-KV / variable KV-partitioning; use a single-SM whole-sequence reduction, or a dual-kernel design with an identical, fixed accumulation order regardless of split count.
- Causal mask + padding mask support; variable-length sequence coverage.
- Keep prefill and decode on a consistent reduction path (coordinate with the KV-cache consistency issue); expose the hooks that issue needs.
- Accumulate softmax statistics in FP32.
- Validate against the #108 harness across batch=1/N, chunked-prefill on/off, padding layouts.

## Out of scope

- Non-standard attention variants (linear / sliding-window / MoE routing) — standard softmax only for this gate.
- Multi-GPU / sequence-parallel attention (WS2).
- FP8 attention.
- Full FlashAttention replacement / maximum throughput in the first deterministic baseline.

## Acceptance criteria

- The same query / sequence produces bitwise-identical (or within #108 tolerance) attention output regardless of batch size, batch position, and chunked-prefill on/off.
- No split-KV path whose split count varies with batch / seq config; the accumulation order is pinned and documented.
- Softmax + LSE outputs are invariant across the sweep; tests cover both fixed-length and variable-length sequences.
- Decode-stage output matches the prefill path for the same effective context (handoff with the KV-cache issue).
- The attention backward path passes the shared gradient-invariance check from the WS1 backward-consistency issue.

## Notes

- Tightly coupled with the KV-cache path-consistency issue — agree on the shared reduction contract early.
- Existing attention kernels (#65 prefix-shared fused attention, #41 Triton FlashAttention fallback) are reference / audit starting points, not the invariant path itself.
- Single-SM whole-sequence is simplest to prove correct but may cost throughput on long sequences; note the tradeoff if choosing the dual-kernel route.

## Planned PRs

- [ ] Deterministic masked-softmax + LSE kernel (fixed max / sum-exp order)
- [ ] Causal + padding mask tests; variable-length coverage
- [ ] Deterministic P x V aggregation (single-SM or fixed-order dual-kernel)
- [ ] Prefill/decode handoff hooks (coordinate with the KV-cache issue)
- [ ] Wire through the #108 harness sweep; benchmark + tradeoff note

### PR DESCRIPTION
**Pull request by EthanZero2Hero**
Closes #147.

This PR adds a Triton standard-softmax attention baseline for WS1 batch-invariant forward validation. The kernel uses a fixed key-order reduction for max, sum-exp/LSE, and P @ V, supports causal and key-padding masks, and is wired into the #108 operator harness as the attention/triton candidate.

## Summary by CodeRabbit

* **New Features**
  * Added a Triton-accelerated attention operator supporting causal and non-causal modes, configurable scaling, key padding masks, and grouped-query attention.
  * Added an API variant that returns both attention output and log-sum-exp (LSE) statistics.

* **Tests**
  * Added CUDA-gated test coverage for forward parity vs the native reference, LSE correctness, padding-layout and batch-position invariance, decode vs. prefill equivalence, kv-cache handoff behavior, and backward/gradient correctness.

* **Chores**
  * Updated attention-specific accuracy tolerances used for kernel verification, and extended the tolerance contract to include the new attention operator.

### CODE REVIEW

File: rl_engine/kernels/ops/triton/attention/standard_attn.py [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+        lse = torch.empty((batch, q_heads, q_len), device=q.device, dtype=torch.float32)
+        block_d = _next_power_of_2(head_dim)
+        grid = (q_len, q_heads, batch)
+        dummy_mask = (
+            key_padding_mask
+            if key_padding_mask is not None
+            else q.new_empty((1,), dtype=torch.bool)
+        )
+
+        _standard_attn_fwd_kernel[grid](
```
Comment (Flink-ddd) [inline rl_engine/kernels/ops/triton/attention/standard_attn.py:189]: Add ctx.mark_non_differentiable(lse) before the return statement when return_lse is True, so PyTorch's autograd engine doesn't expect a gradient for the LSE tensor (since your custom backward explicitly ignores grad_lse).

File: rl_engine/kernels/gtest/operator_specs.py [hunk 2]
```diff
@@ -32,6 +32,20 @@ def _load_object(path: str) -> Any:
 
 
 OP_SPECS = {
+    "attention": OperatorSpec(
+        name="attention",
+        op_class="reduction",
```
Comment (Flink-ddd) [inline rl_engine/kernels/gtest/operator_specs.py]: Change op_class="reduction" to op_class="attention" to ensure the test harness maps this Triton candidate to the correct attention tolerance contract established in PR #240.

**Review by Flink-ddd — CHANGES_REQUESTED** [review]
Left a few comments, otherwise LGTM!

File: rl_engine/kernels/ops/triton/attention/standard_attn.py (hunk 1 above)
Comment (EthanZero2Hero) [reply rl_engine/kernels/ops/triton/attention/standard_attn.py:189]: Thanks, fixed.

I added `ctx.mark_non_differentiable(lse)` before returning `(out, lse)` when `return_lse=True`, so autograd will not expect or propagate a gradient for the LSE tensor. I also added a regression test that verifies `lse.requires_grad is False` while gradients for q/k/v still work through the attention output.

File: rl_engine/kernels/gtest/operator_specs.py (hunk 2 above)
Comment (EthanZero2Hero) [reply rl_engine/kernels/gtest/operator_specs.py]: Thanks, addressed.

Changed the attention operator spec from `op_class="reduction"` to `op_class="attention"` so the #108 harness uses the attention-specific tolerance contract.

I also added the `attention` tolerance entry locally, matching the contract from PR #240, so this PR can run the harness cleanly on the current base.

File: tests/test_triton_batch_invariant_attention.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+def test_triton_attention_decode_matches_prefill_suffix_context():
+    dtype = torch.bfloat16
+    q_full, k_full, v_full = _qkv(1, 12, 12, dtype=dtype, seed=6)
+    op = TritonBatchInvariantAttentionOp()
+
+    prefill = op(q_full, k_full, v_full, causal=True)
+    decode = op(q_full[:, :, -1:], k_full, v_full, causal=True)
+
+    assert torch.equal(prefill[:, :, -1:], decode)
+
```
Comment (maxiaosong1124) [inline tests/test_triton_batch_invariant_attention.py:259]: Please add coverage equivalent to the CUDA tests for:
  - full prefill vs sequence-dimension chunked prefill with several chunk sizes;
  - chunked-prefill LSE invariance;
  - chunked prefill with key_padding_mask;
  - KV-cache handoff with S_new=1 and S_new>1;
  - representative MHA/GQA configurations.

  The comparisons for equivalent contexts should use torch.equal where the
  batch-invariance contract requires bitwise identity.

File: tests/test_triton_batch_invariant_attention.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    out.backward(dy)
+    return out.detach(), q_req.grad.detach(), k_req.grad.detach(), v_req.grad.detach()
+
+
+@requires_cuda
+@pytest.mark.parametrize("causal", [False, True])
+@pytest.mark.parametrize(
+    "dtype, atol, rtol", [(torch.float16, 3e-3, 3e-3), (torch.bfloat16, 5e-2, 2e-2)]
+)
+def test_triton_attention_matches_native_forward(causal, dtype, atol, rtol):
```
Comment (maxiaosong1124) [inline tests/test_triton_batch_invariant_attention.py:68]: Please add an all-False key_padding_mask row and verify:
  - output is finite and exactly zero;
  - LSE is -inf;
  - behavior matches the NativeAttentionOp convention;
  - backward produces finite gradients and no contribution from the fully masked
    row.

**Review by maxiaosong1124 — COMMENTED** [review]
Left some comments,please resolve these problems, thanks!

File: tests/test_triton_batch_invariant_attention.py (hunk 3 above)
Comment (EthanZero2Hero) [reply tests/test_triton_batch_invariant_attention.py:259]: Thanks, added Triton coverage equivalent to the CUDA batch-invariance tests.

This update adds:
- full prefill vs sequence-dimension chunked prefill across several chunk sizes;
- chunked-prefill LSE invariance;
- chunked prefill with `key_padding_mask`;
- KV-cache handoff coverage for both `S_new=1` and `S_new>1`;
- representative MHA/GQA configurations.

For equivalent contexts, the tests now use `torch.equal` for both output and LSE to enforce the bitwise batch-invariance contract.

File: tests/test_triton_batch_invariant_attention.py (hunk 4 above)
Comment (EthanZero2Hero) [reply tests/test_triton_batch_invariant_attention.py:68]: Thanks, added an all-False `key_padding_mask` row test.

The test verifies:
- the fully masked row output is finite and exactly zero;
- LSE is `-inf`;
- the behavior matches `NativeAttentionOp`;
- backward produces finite gradients;
- the fully masked row contributes exactly zero gradients for q/k/v.

**Review by Flink-ddd — APPROVED** [review]
LGTM now, Thank you for update.

**Review by maxiaosong1124 — APPROVED** [review]
LGTM now! thanks for your update!
